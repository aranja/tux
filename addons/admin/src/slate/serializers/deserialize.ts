import { Value, State } from 'slate'
import Plain from 'slate-plain-serializer'
import { Html } from './Html'
import { Format } from './Format'

function deserialize(value: any, format: Format = 'raw'): State | null {
  // Don't try to deserialize null.
  if (value == null) {
    return null
  }

  // Always handle raw values.
  // Make it easier to migrate between raw and html.
  if (typeof value === 'object' && value.nodes) {
    return Value.fromJSON(value)
  }

  if (typeof value !== 'string') {
    return null
  }

  switch (format) {
    case 'plain':
      return Plain.deserialize(value)
    // If value is string and format raw,
    case 'raw':
    case 'html':
      return Html.deserialize(value)
    default:
      throw new Error(`Don't know how to deserialize ${format}`)
  }
}

export default deserialize
