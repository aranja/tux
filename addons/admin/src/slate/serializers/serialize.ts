import { State } from 'slate'
import Plain from 'slate-plain-serializer'
import { Html } from './Html'
import { Format } from './Format'

function serialize(state: State, format: Format = 'raw'): any {
  switch (format) {
    case 'raw':
      return state.toJSON()
    case 'html':
      return Html.serialize(state)
    case 'plain':
      return Plain.serialize(state)
    default:
      throw new Error(`Don't know how to deserialize ${format}`)
  }
}

export default serialize
