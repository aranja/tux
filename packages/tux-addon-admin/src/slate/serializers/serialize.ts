import { Plain, Raw, State } from 'slate'
import { Html } from './Html'
import { Format } from './Format'

function serialize(state: State, format: Format = 'raw'): any {
  switch (format) {
    case 'raw':
      return Raw.serialize(state, { terse: true })
    case 'html':
      return Html.serialize(state)
    case 'plain':
      return Plain.serialize(state)
    default:
      throw new Error(`Don't know how to deserialize ${format}`)
  }
}

export default serialize
