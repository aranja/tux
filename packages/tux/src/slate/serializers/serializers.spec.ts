import deserialize from './deserialize'
import serialize from './serialize'

const testHtml = `<p>Hello <em>italic</em> <strong>bold</strong> <u>underlined</u> <a href="foo">link</a> <code>code</code></p><ul><li>List</li></ul><ol><li>Ordered list</li></ol><blockquote><code>Fancy code</code></blockquote><h1>H1</h1><h2>H2</h2><h3>H3</h3><h4>H4</h4><h5>H5</h5><h6>H6</h6>`

describe('serializers', () => {
  it('safely roundtrips html', () => {
    const state = deserialize(testHtml, 'html')
    const newHtml = serialize(state, 'html')
    expect(newHtml).toEqual(testHtml)
  })

  it('safely roundtrips plain', () => {
    const oldPlain = `
    Test test.
    
    Ok.
    `

    const state = deserialize(oldPlain, 'plain')
    const newPlain = serialize(state, 'plain')
    expect(newPlain).toEqual(oldPlain)
  })

  it('safely roundtrips html and raw', () => {
    let state = deserialize(testHtml, 'html')
    const raw = serialize(state, 'raw')
    expect(raw).toMatchSnapshot()

    state = deserialize(raw, 'raw')
    const newHtml = serialize(state, 'html')
    expect(newHtml).toEqual(testHtml)
  })
})
