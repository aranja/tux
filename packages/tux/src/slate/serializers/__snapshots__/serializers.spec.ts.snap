exports[`serializers safely roundtrips html and raw 1`] = `
Object {
  "nodes": Array [
    Object {
      "kind": "block",
      "nodes": Array [
        Object {
          "kind": "text",
          "ranges": Array [
            Object {
              "text": "Hello ",
            },
            Object {
              "marks": Array [
                Object {
                  "type": "italic",
                },
              ],
              "text": "italic",
            },
            Object {
              "text": " ",
            },
            Object {
              "marks": Array [
                Object {
                  "type": "bold",
                },
              ],
              "text": "bold",
            },
            Object {
              "text": " ",
            },
            Object {
              "marks": Array [
                Object {
                  "type": "underline",
                },
              ],
              "text": "underlined",
            },
            Object {
              "text": " ",
            },
          ],
        },
        Object {
          "data": Object {
            "href": "foo",
          },
          "kind": "inline",
          "nodes": Array [
            Object {
              "kind": "text",
              "text": "link",
            },
          ],
          "type": "link",
        },
        Object {
          "kind": "text",
          "ranges": Array [
            Object {
              "text": " ",
            },
            Object {
              "marks": Array [
                Object {
                  "type": "code",
                },
              ],
              "text": "code",
            },
          ],
        },
      ],
      "type": "paragraph",
    },
    Object {
      "kind": "block",
      "nodes": Array [
        Object {
          "kind": "block",
          "nodes": Array [
            Object {
              "kind": "text",
              "text": "List",
            },
          ],
          "type": "list-item",
        },
      ],
      "type": "bulleted-list",
    },
    Object {
      "kind": "block",
      "nodes": Array [
        Object {
          "kind": "block",
          "nodes": Array [
            Object {
              "kind": "text",
              "text": "Ordered list",
            },
          ],
          "type": "list-item",
        },
      ],
      "type": "numbered-list",
    },
    Object {
      "kind": "block",
      "nodes": Array [
        Object {
          "kind": "text",
          "ranges": Array [
            Object {
              "marks": Array [
                Object {
                  "type": "code",
                },
              ],
              "text": "Fancy code",
            },
          ],
        },
      ],
      "type": "quote",
    },
    Object {
      "kind": "block",
      "nodes": Array [
        Object {
          "kind": "text",
          "text": "H1",
        },
      ],
      "type": "heading-one",
    },
    Object {
      "kind": "block",
      "nodes": Array [
        Object {
          "kind": "text",
          "text": "H2",
        },
      ],
      "type": "heading-two",
    },
    Object {
      "kind": "block",
      "nodes": Array [
        Object {
          "kind": "text",
          "text": "H3",
        },
      ],
      "type": "heading-three",
    },
    Object {
      "kind": "block",
      "nodes": Array [
        Object {
          "kind": "text",
          "text": "H4",
        },
      ],
      "type": "heading-four",
    },
    Object {
      "kind": "block",
      "nodes": Array [
        Object {
          "kind": "text",
          "text": "H5",
        },
      ],
      "type": "heading-five",
    },
    Object {
      "kind": "block",
      "nodes": Array [
        Object {
          "kind": "text",
          "text": "H6",
        },
      ],
      "type": "heading-six",
    },
  ],
}
`;
